internal/app/machined/pkg/runtime/v1alpha1/bootloader/grub/grub.go

```diff
@@ -0,0 +1,192 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package grub
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	goruntime "runtime"
+	"strings"
+	"text/template"
+
+	"github.com/talos-systems/talos/internal/app/machined/pkg/runtime"
+	"github.com/talos-systems/talos/pkg/blockdevice/probe"
+	"github.com/talos-systems/talos/pkg/blockdevice/util"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+)
+
+// Cfg reprsents the cfg file.
+type Cfg struct {
+	Default  string
+	Fallback string
+	Labels   []*Label
+}
+
+// Label reprsents a label in the cfg file.
+type Label struct {
+	Root   string
+	Kernel string
+	Initrd string
+	Append string
+}
+
+const grubCfgTpl = `set default="{{ .Default }}"
+{{ with .Fallback -}}
+set fallback="{{ . }}"
+{{- end }}
+set timeout=0
+
+{{ range $label := .Labels -}}
+menuentry "{{ $label.Root }}" {
+  linux {{ $label.Kernel }} {{ $label.Append }}
+  initrd {{ $label.Initrd }}
+}
+{{- end }}
+`
+
+// Grub represents the grub bootloader.
+type Grub struct{}
+
+// Labels implements the Bootloader interface.
+func (g *Grub) Labels() (current, next string, err error) {
+	var b []byte
+
+	if b, err = ioutil.ReadFile(GrubConfig); err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			next = BootA
+
+			return current, next, nil
+		}
+
+		return "", "", err
+	}
+
+	re := regexp.MustCompile(`^set default="(.*)"`)
+	matches := re.FindAllSubmatch(b, -1)
+
+	if len(matches) != 1 {
+		return "", "", fmt.Errorf("failed to find default")
+	}
+
+	if len(matches[0]) != 2 {
+		log.Printf("%+v", matches[0])
+		return "", "", fmt.Errorf("expected 2 matches, got %d", len(matches[0]))
+	}
+
+	current = string(matches[0][1])
+	switch current {
+	case BootA:
+		next = BootB
+	case BootB:
+		next = BootA
+	default:
+		return "", "", fmt.Errorf("unknown grub menuentry: %q", current)
+	}
+
+	return current, next, err
+}
+
+// Install implements the Bootloader interface. It sets up grub with the
+// specified kernel parameters.
+//
+// nolint: gocyclo
+func (g *Grub) Install(fallback string, config interface{}, sequence runtime.Sequence, bootPartitionFound bool) (err error) {
+	grubcfg, ok := config.(*Cfg)
+	if !ok {
+		return errors.New("expected a grub config")
+	}
+
+	if err = writeCfg(GrubConfig, grubcfg); err != nil {
+		return err
+	}
+
+	dev, err := probe.GetDevWithFileSystemLabel(constants.BootPartitionLabel)
+	if err != nil {
+		return fmt.Errorf("failed to probe boot partition: %w", err)
+	}
+
+	// nolint: errcheck
+	defer dev.Close()
+
+	blk, err := util.DevnameFromPartname(dev.Path)
+	if err != nil {
+		return err
+	}
+
+	loopDevice := strings.HasPrefix(blk, "loop")
+
+	blk = fmt.Sprintf("/dev/%s", blk)
+
+	// TODO: run twice
```

Comment:
I forgot to remove this `TODO:`, we can remove it now (this was to run twice on x86 for bios/efi)